### ISSUE TITLE
Tokens change ownership again after restarting replacing node in replace_first_boot_test

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylla#5906]
Regression seen since https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/383/testReport/replace_address_test/TestReplaceAddress/replace_first_boot_test/
Scylla version 2816404f575545ac40933bbbd8621abade2f6c9f
```
Traceback (most recent call last):
  File "/usr/lib64/python3.7/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib64/python3.7/unittest/case.py", line 645, in run
    testMethod()
  File "/jenkins/workspace/scylla-master/dtest-release/scylla-dtest/replace_address_test.py", line 240, in replace_first_boot_test
    self.assertEqual(len(movedTokensList), numNodes)
  File "/usr/lib64/python3.7/unittest/case.py", line 869, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib64/python3.7/unittest/case.py", line 862, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 512 != 256
```

Test replaces node3 with node4 while node3 is stopped.
It counts the number of tokens that changed ownership by the number of "Token .* changing ownership from node3.address() to node4.address()" messages in node4's log (256 tokens in this test).

It then restarts node3.

Note: at this point I saw corresponding 256 "changing ownership" messages on node3 log when trying to reproduce locally - which is suspicious and could be related to https://github.com/scylladb/scylla/issues/5523. The test actually passes in this case, maybe it should not, once #5523 is fixed.

In the failure case, node4 is then restarted and the test counts again the number pf "changing ownership" message in node4's log (starting from the beginning of the log) and the number should remain the same.  However, quite frequently the messages are repeated and the count is twice as expected (512 vs. 256).

@asias: does this indicate that node4, when rebooted, still thinks the tokens are owned by node3 and moved to node4 (again)?  What are the ramifications of this issue? Can it load to data corruption? For example, if the client accesses the data on node3 in the transition time with `CL=LOCAL_ONE`?

**Comment by bhalevy** [comment]
Most recently hit in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/400/testReport/replace_address_test/TestReplaceAddress/replace_first_boot_test/
Logs are in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/400/artifact/logs-release.2/1582685714649_replace_address_test.TestReplaceAddress.replace_first_boot_test/

**Comment by asias** [comment]
I will look into it. It is probably something wrong with how we log. The replacing operation itself should be fine. Regarding the restarting the replaced node again issue, it has lower priority though.

**Comment by asias** [comment]
The test is problematic.
```
222         # check that restarting node 3 doesn't work
223         # FIXME: https://github.com/scylladb/scylla/issues/5523 is fixed
224         # need to verify that the node doesn't start listening
225         debug("Try to restart node 3 (should fail)")
226         node3.start(no_wait=True)
227         checkCollision = node1.grep_log("between .*"+self.cluster.get_node_ip(3)+" and .*"+self.cluster.get_node_ip(4)+"; .*"+self.cluster.get_node_ip(4)+" is the new owner")
228         debug(checkCollision)
229         self.assertEqual(len(checkCollision), 1)
230 
231         # restart node4 (if error's might have to change num_tokens)
232         node4.stop(gently=False)
233         node4.start(wait_for_binary_proto=True, wait_other_notice=False)
234 
235         debug("Verifying querying works again.")
236         finalData = list(session.execute(query))
237         self.assertCountEqual(initialData, finalData)
238 
239         # we redo this check because restarting node should not result in tokens being moved again, ie number should be same
240         debug("Verifying tokens migrated sucessfully")
241         movedTokensList = node4.grep_log("Token .* changing ownership from .*"+self.cluster.get_node_ip(3)+" to .*"+self.cluster.get_node_ip(4))
242         debug(movedTokensList[0])
243         self.assertEqual(len(movedTokensList), numNodes)
```

The check at line 227 will pass when n4 is finished replacing. It has nothing to do with the restart of the replaced node here. The restart of replaced node n3 will not faill at all. 

At line 232, we restart n4 immediately, note at line 226, we did not wait fro n3's restart. If we really want to see what happens if n3 restarts, we should wait for it to finish restaring. 

The check at line 243 is problematic, because nothing prevents we get the log changing ownership from n3 to n4 again, after restart of n3 and n4.  

After n3's restart at line 226, n3 will be in gossip, when n4 restarts, if n3's state is present at n4, n4 then goes to normal, we sill see n4 logs that ownership is changed from n3 to n4. 

In short, I think we should remove the restart of n3 at line 226 in the test, it is an operation error and we can not detect it yet.

**Comment by asias** [comment]
I propose a patch to dtest like this:

```
--- a/replace_address_test.py
+++ b/replace_address_test.py
@@ -219,15 +219,16 @@ class TestReplaceAddress(Tester):
         debug(movedTokensList[0])
         self.assertEqual(len(movedTokensList), numNodes)
 
-        # check that restarting node 3 doesn't work
-        # FIXME: https://github.com/scylladb/scylla/issues/5523 is fixed
-        # need to verify that the node doesn't start listening
-        debug("Try to restart node 3 (should fail)")
-        node3.start(no_wait=True)
         checkCollision = node1.grep_log("between .*"+self.cluster.get_node_ip(3)+" and .*"+self.cluster.get_node_ip(4)+"; .*"+self.cluster.get_node_ip(4)+" is the new owner")
         debug(checkCollision)
         self.assertEqual(len(checkCollision), 1)
 
+        # check that restarting node 3 doesn't work
+        # FIXME: https://github.com/scylladb/scylla/issues/5523 is fixed
+        # need to verify that the node doesn't start listening
+        # debug("Try to restart node 3 (should fail)")
+        # node3.start(no_wait=True)
+
         # restart node4 (if error's might have to change num_tokens)
         node4.stop(gently=False)
         node4.start(wait_for_binary_proto=True, wait_other_notice=False)
```

**Comment by bhalevy** [comment]
> I propose a patch to dtest like this:

The change LGTM, can you please send a pull request to scylla-dtest?

**Comment by asias** [comment]
Done: https://github.com/scylladb/scylla-dtest/pull/1340

**Comment by asias** [comment]
The patch is merged: scylladb/scylla-dtest#1340. Closing it now. Please reopen if needed.
